### PR TITLE
[dtensor] supports uneven sharding

### PR DESCRIPTION
**Pull request by wanchaol**
This PR supports uneven sharding, we remove the constraint that it
must be sharding evenly as most of the tensors/parameters on one
tensor dimension might be not divisible to world_size

Concretely this PR:
1. Add the capability for uneven size sharding in distribute_tensor
2. Add the capability to redistribute a uneven sized tensor
3. standardize most sharding calculation to `Shard` spec directly
4. refactor DeviceMesh to support more flexible sharding in the future

TODO:
1. make `from_local` to accept uneven sharding
2. make sure empty shards be ignored during computation
3. more tests for corner cases, i.e. stride testing
4. dynamic shapes for tracing

### CODE REVIEW

File: spmd/tensor/device_mesh.py
Comment (awgu) [inline spmd/tensor/device_mesh.py:470]: Curious: Does `reduce_scatter()` need to use `CommTensor` as well?

File: spmd/tensor/device_mesh.py
Comment (awgu) [inline spmd/tensor/device_mesh.py:474]: This is probably premature optimization but IIUC:

For `mesh_dim` equal to `k` and assuming equal-sized size-`n`-byte tensors in `tensor_list` , the current gloo mesh reduce-scatter allocates `(k-1) * n` bytes (via `clone()`); copies `k * n` bytes (`n` via `copy_()` and `(k-1) * n` via `clone()`); and calls `k` all-reduces.

I think by increasing the memory usage slightly, you may decrease the number of all-reduces to just `1`: You allocate `k * n` bytes at once in a single tensor (i.e. `torch.empty()`), copy `k * n` bytes into tensor (i.e. the `input_list` contents), call `1` all-reduce, and copy `n` bytes into `output`. This allocates the same number of bytes and requires copying `n` additional bytes but makes the number of all-reduces a constant `1`.

I am not sure how the single `n * k`-byte tensor allocation compares to `k` separate `n`-byte allocations in terms of memory fragmentation,. but I hypothesize that for small and medium sizes (which DT may target), the `n * k` byte allocation will be better.

**Comment by kumpera** [comment]
With this change DCP will require DT to expose the bounding box of the local shard as it's becoming increasingly trickier to find it. cc @wz337

File: spmd/tensor/device_mesh.py
Comment (wanchaol) [reply spmd/tensor/device_mesh.py:474]: Thanks for the suggestion! Yeah this looks better, will update.

**Comment by wanchaol** [comment]
> With this change DCP will require DT to expose the bounding box of the local shard as it's becoming increasingly trickier to find it. cc @wz337

Done this in a separate PR https://github.com/pytorch/tau/pull/550, after that landed, this PR will add more uneven testing cases for shard offsets.

File: spmd/tensor/utils.py
Comment (aazzolini) [inline spmd/tensor/utils.py:32]: nit: inline return?

File: spmd/tensor/dispatch.py
Comment (aazzolini) [inline spmd/tensor/dispatch.py:123]: what is the reason for moving this?

File: spmd/tensor/ops/common_rules.py
Comment (aazzolini) [inline spmd/tensor/ops/common_rules.py:330]: nit: is there a reason to keep dims_list around before normalization? is that used below? if not , better remove to avoid confusion

File: spmd/tensor/ops/tensor_ops.py
Comment (aazzolini) [inline spmd/tensor/ops/tensor_ops.py:227]: nice catch! thanks for the fix.

File: spmd/tensor/ops/tensor_ops.py
Comment (aazzolini) [inline spmd/tensor/ops/tensor_ops.py:447]: thanks for the fix!

File: spmd/tensor/ops/tp_sharding_ops.py
Comment (aazzolini) [inline spmd/tensor/ops/tp_sharding_ops.py:22]: Was this a bug, previously?

File: spmd/tensor/ops/view_ops.py
Comment (aazzolini) [inline spmd/tensor/ops/view_ops.py:407]: nice catch!

File: spmd/tensor/ops/view_ops.py
Comment (aazzolini) [inline spmd/tensor/ops/view_ops.py:232]: Would Sequence work instead of Tuple | List?

File: spmd/tensor/parallel/_view_with_dim_change.py
Comment (aazzolini) [inline spmd/tensor/parallel/_view_with_dim_change.py:40]: Is logical shape the same as "global" shape?

File: spmd/tensor/__init__.py
Comment (aazzolini) [inline spmd/tensor/__init__.py:85]: Hadn't we added "contiguous" to DeviceMesh impl directly?

File: spmd/tensor/api.py
Comment (aazzolini) [inline spmd/tensor/api.py:85]: Is there a case where we wouldn't want to make it contiguous before broadcast?

File: spmd/tensor/api.py
Comment (aazzolini) [inline spmd/tensor/api.py:91]: Sounds risky -- is run_check=True by default?

File: spmd/tensor/api.py
Comment (aazzolini) [inline spmd/tensor/api.py:181]: How does this actually work? Are we creating a tensor with custom strides? In case the tensor is not contiguous, what happens in practice?

Comment (aazzolini) [reply spmd/tensor/api.py:181]: For uneven sharding, would we need to keep uneven strides?

File: spmd/tensor/api.py
Comment (aazzolini) [inline spmd/tensor/api.py:328]: My gut feeling is that we should leave that to the compiler.

File: spmd/tensor/ops/utils.py
Comment (aazzolini) [inline spmd/tensor/ops/utils.py:63]: could we add quick docstring?

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:55]: Is this assertion necessary now that we support uneven shards? E.g. if we have 4 devices but a tensor of size 3, we could support it by having a device holding an empty tensor (or a 1 element tensor in case with_padding=True).

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:69]: are there cases where we don't expect contiguous to be true but still want padding? Also is there a reason not leave contiguous for the comm backend to do in case needed?

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:79]: what does this lower to in terms of aten ops? Does it bake in the static tensor shape when doing tracing? This would be an issue if we would like to support dynamic shapes on tracing mode.

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:84]: Does this bake in a static shape when tracing? This would be a problem if we want to support dynamic shapes.

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:94]: turn this into a proper docstring?

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:97]: Same question about why not supporting size_on_dim < num_chunks since empty tensors should work ok ?

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:92]: in case of tracing dynamic shapes would we be getting SymInt as input here during tracing instead of ints?

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:113]: Make this into a proper docstring?

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:134]: turn into docstring

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:132]: It feels a bit weird that this is a member of Shard class and that it takes a _non_ DTensor. 
1) does this need to be a member of Shard?
2) what really are we expecting/allowing as "tensor" argument? Is it supposed to be one _shard_ of a DTensor? i can see that but this would at most be ok as a private API. Seems too confusing/risky for end-user to use.

File: spmd/tensor/placement_types.py
Comment (aazzolini) [inline spmd/tensor/placement_types.py:153]: Same comment here -- this feels a bit weird as a member of Shard, specially as a public API. I'd be more comfortable with an _

**Review by aazzolini — COMMENTED** [review]
This is a great piece of work, thanks for doing this!
My main concern is whether we should expose these APIs as public under Shard. I'd feel more comfortable making them available only to the implementation of higher level methods.

File: spmd/tensor/placement_types.py
Comment (wanchaol) [reply spmd/tensor/placement_types.py:79]: It adds an additional node like below:

`torch.ops.aten.constant_pad_nd.default(slice_2, [0, 0, 0, 1], 0.0)`

IF we want to do dynamic shapes, my reading with dynamic shapes is that tracing would generate symbolic shapes instead of static constant shapes, so we should be fine, but we would have to test how it looks like when symbolic shape tracing works.

File: spmd/tensor/placement_types.py
Comment (wanchaol) [reply spmd/tensor/placement_types.py:92]: Yeah i think so, but I think we don't trace sharding propagation directly? We trace user model, and expand user model with dtensor expansion, which we can still run sharding propagation on python?

File: spmd/tensor/placement_types.py
Comment (wanchaol) [reply spmd/tensor/placement_types.py]: The reason I keep all the sharding related logic into `Shard` directly is that the way to do sharding `tensor_split`, and padding with `tensor_split` are specific to `tensor_split`, if we want to do another type of sharding, it's easy to just extend the `Placement` directly and implement some sharding logic that is required by `distribute_tensor` and `redistribute`, (i.e. for interleaved sharding we can easily implement a interleaved placement that would work). 

Agree that this is actually all internal logic and we should not expose this to end-user, I'll make the methods be all private.

File: spmd/tensor/parallel/_view_with_dim_change.py
Comment (aazzolini) [reply spmd/tensor/parallel/_view_with_dim_change.py:40]: global makes more sense to me but i'm okay either way, as long as we're consistent.

File: spmd/tensor/__init__.py
Comment (aazzolini) [reply spmd/tensor/__init__.py:85]: I see, in-place support really complicates things.

File: spmd/tensor/api.py
Comment (aazzolini) [reply spmd/tensor/api.py:85]: Makes sense, given support for in-place operations.

File: spmd/tensor/placement_types.py
Comment (aazzolini) [reply spmd/tensor/placement_types.py:55]: Got it so you did try it out and it didn't work? (I know that c10d segfaults when passed an empty tensor ... but here we're padding, so that shouldn't be a problem).

**Review by aazzolini — APPROVED** [review]
Thanks for the changes and for the clarifications. We can follow up on separate diffs re. 1) dynamic shapes; 2) tensors smaller than number of ranks.

**Pull request merged**
